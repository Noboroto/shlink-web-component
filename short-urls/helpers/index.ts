import { isNil } from 'ramda';
import type { OptionalString } from '../../utils/helpers';
import type { ShortUrlCreationSettings } from '../../utils/settings';
import { DEFAULT_DOMAIN } from '../../visits/reducers/domainVisits';
import type { ShortUrl, ShortUrlData } from '../data';

export const shortUrlMatches = (shortUrl: ShortUrl, shortCode: string, domain: OptionalString): boolean => {
  if (isNil(domain)) {
    return shortUrl.shortCode === shortCode && !shortUrl.domain;
  }

  return shortUrl.shortCode === shortCode && shortUrl.domain === domain;
};

export const domainMatches = (shortUrl: ShortUrl, domain: string): boolean => {
  if (!shortUrl.domain && domain === DEFAULT_DOMAIN) {
    return true;
  }

  return shortUrl.domain === domain;
};

export const shortUrlDataFromShortUrl = (shortUrl?: ShortUrl, settings?: ShortUrlCreationSettings): ShortUrlData => {
  const validateUrl = settings?.validateUrls ?? false;

  if (!shortUrl) {
    return { longUrl: '', validateUrl };
  }

  return {
    longUrl: shortUrl.longUrl,
    tags: shortUrl.tags,
    title: shortUrl.title ?? undefined,
    domain: shortUrl.domain ?? undefined,
    validSince: shortUrl.meta.validSince ?? undefined,
    validUntil: shortUrl.meta.validUntil ?? undefined,
    maxVisits: shortUrl.meta.maxVisits ?? undefined,
    crawlable: shortUrl.crawlable,
    forwardQuery: shortUrl.forwardQuery,
    deviceLongUrls: shortUrl.deviceLongUrls,
    validateUrl,
  };
};

const MULTI_SEGMENT_SEPARATOR = '__';

export const urlEncodeShortCode = (shortCode: string): string => shortCode.replaceAll('/', MULTI_SEGMENT_SEPARATOR);

export const urlDecodeShortCode = (shortCode: string): string => shortCode.replaceAll(MULTI_SEGMENT_SEPARATOR, '/');
