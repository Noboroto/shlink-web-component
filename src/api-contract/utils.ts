import type {
  InvalidArgumentError,
  InvalidShortUrlDeletion,
  ProblemDetailsError,
} from '.';
import { ErrorType } from '.';

export const isInvalidArgumentError = (
  error?: ProblemDetailsError
): error is InvalidArgumentError => error?.type === ErrorType.INVALID_ARGUMENT;

export const isInvalidDeletionError = (
  error?: ProblemDetailsError
): error is InvalidShortUrlDeletion =>
  error?.type === ErrorType.INVALID_SHORT_URL_DELETION;

const isProblemDetails = (e: unknown): e is ProblemDetailsError =>
  !!e &&
  typeof e === 'object' &&
  ['type', 'detail', 'title', 'status'].every((prop) => prop in e);

export const parseApiError = (e: unknown): ProblemDetailsError | undefined =>
  isProblemDetails(e) ? e : undefined;

export const isErrorAction = (action: any): boolean => 'error' in action;
