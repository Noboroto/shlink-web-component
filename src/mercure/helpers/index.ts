import { EventSourcePolyfill as EventSource } from 'event-source-polyfill';
import type { MercureInfo } from '../reducers/mercureInfo';

export const bindToMercureTopic = <T>(
  mercureInfo: MercureInfo,
  topics: string[],
  onMessage: (message: T) => void,
  onTokenExpired: () => void
) => {
  const { mercureHubUrl, token, loading, error } = mercureInfo;

  if (loading || error || !mercureHubUrl) {
    return undefined;
  }

  const onEventSourceMessage = ({ data }: { data: string }) =>
    onMessage(JSON.parse(data) as T);
  const onEventSourceError = ({ status }: { status: number }) =>
    status === 401 && onTokenExpired();

  const subscriptions = topics.map((topic) => {
    const hubUrl = new URL(mercureHubUrl);

    hubUrl.searchParams.append('topic', topic);
    const es = new EventSource(hubUrl, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });

    es.onmessage = onEventSourceMessage;
    es.onerror = onEventSourceError;

    return es;
  });

  return () => subscriptions.forEach((es) => es.close());
};
