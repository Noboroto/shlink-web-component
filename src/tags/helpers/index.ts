const ONE_OR_MORE_SPACES_REGEX = /\s+/g;

export const normalizeTag = (tag: string) => tag.trim().toLowerCase().replace(ONE_OR_MORE_SPACES_REGEX, '-');
