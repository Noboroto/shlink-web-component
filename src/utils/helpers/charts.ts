import type { ActiveElement, ChartEvent, ChartType, TooltipItem } from 'chart.js';
import { prettify } from './numbers';

export const pointerOnHover = ({ native }: ChartEvent, [firstElement]: ActiveElement[]) => {
  if (!native?.target) {
    return;
  }

  const canvas = native.target as HTMLCanvasElement;

  canvas.style.cursor = firstElement ? 'pointer' : 'default';
};

export const renderChartLabel = ({ dataset, raw }: TooltipItem<ChartType>) => `${dataset.label}: ${prettify(`${raw}`)}`;

export const renderPieChartLabel = ({ label, raw }: TooltipItem<ChartType>) => `${label}: ${prettify(`${raw}`)}`;
