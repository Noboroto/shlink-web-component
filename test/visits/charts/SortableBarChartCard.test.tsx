import { range } from '@shlinkio/data-manipulation';
import { screen } from '@testing-library/react';
import type { ReactNode } from 'react';
import { rangeOf } from '../../../src/utils/helpers';
import { SortableBarChartCard } from '../../../src/visits/charts/SortableBarChartCard';
import type { Stats } from '../../../src/visits/types';
import { checkAccessibility } from '../../__helpers__/accessibility';
import { renderWithEvents } from '../../__helpers__/setUpTest';

describe('<SortableBarChartCard />', () => {
  const sortingItems = {
    name: 'Name',
    amount: 'Amount',
  };
  const stats = {
    Foo: 100,
    Bar: 50,
  };
  const setUp = (withPagination = false, extraStats = {}, extra?: (foo?: string[]) => ReactNode) => renderWithEvents(
    <SortableBarChartCard
      title="Foo"
      stats={{ ...stats, ...extraStats }}
      sortingItems={sortingItems}
      withPagination={withPagination}
      extraHeaderContent={extra}
    />,
  );

  it('passes a11y checks', () => checkAccessibility(setUp()));

  it('renders stats unchanged when no ordering is set', () => {
    const { container } = setUp();

    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  it.each([
    ['Name', 1],
    ['Amount', 1],
    ['Name', 2],
    ['Amount', 2],
  ])('renders properly ordered stats when ordering is set', async (name, clicks) => {
    const { user } = setUp();

    await user.click(screen.getByRole('button'));
    await Promise.all(rangeOf(clicks, async () => user.click(screen.getByRole('menuitem', { name }))));

    expect(screen.getByRole('document')).toMatchSnapshot();
  });

  it.each([
    [0],
    [1],
    [2],
    [3],
  ])('renders properly paginated stats when pagination is set', async (itemIndex) => {
    const { user } = setUp(true, range(1, 159).reduce<Stats>((accum, value) => {
      accum[`key_${value}`] = value;
      return accum;
    }, {}));

    await user.click(screen.getAllByRole('button')[1]);
    await user.click(screen.getAllByRole('menuitem')[itemIndex]);

    expect(screen.getByRole('document')).toMatchSnapshot();
  });

  it('renders extra header content', () => {
    setUp(false, {}, () => (
      <span>
        <span className="foo-span">Foo in header</span>
        <span className="bar-span">Bar in header</span>
      </span>
    ));

    expect(screen.getByText('Foo in header')).toHaveClass('foo-span');
    expect(screen.getByText('Bar in header')).toHaveClass('bar-span');
  });
});
