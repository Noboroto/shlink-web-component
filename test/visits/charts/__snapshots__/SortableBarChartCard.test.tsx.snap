// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SortableBarChartCard /> > renders highlighted and prev stats when provided 1`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="disabled dropdown-item"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <div
      class="recharts-responsive-container"
      style="width: 100%; height: 300px; min-width: 0;"
    />
  </div>
</div>
`;

exports[`<SortableBarChartCard /> > renders properly ordered stats when ordering is set 1`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-up-wide-short ordering-dropdown__sort-icon"
              data-icon="arrow-up-wide-short"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M151.6 42.4C145.5 35.8 137 32 128 32s-17.5 3.8-23.6 10.4l-88 96c-11.9 13-11.1 33.3 2 45.2s33.3 11.1 45.2-2L96 146.3V448c0 17.7 14.3 32 32 32s32-14.3 32-32V146.3l32.4 35.4c11.9 13 32.2 13.9 45.2 2s13.9-32.2 2-45.2l-88-96zM320 480h32c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32zm0-128h96c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32zm0-128H480c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32zm0-128H544c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32z"
                fill="currentColor"
              />
            </svg>
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <div
      class="recharts-responsive-container"
      style="width: 100%; height: 300px; min-width: 0;"
    />
  </div>
</div>
`;

exports[`<SortableBarChartCard /> > renders properly ordered stats when ordering is set 2`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
          </button>
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-up-wide-short ordering-dropdown__sort-icon"
              data-icon="arrow-up-wide-short"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M151.6 42.4C145.5 35.8 137 32 128 32s-17.5 3.8-23.6 10.4l-88 96c-11.9 13-11.1 33.3 2 45.2s33.3 11.1 45.2-2L96 146.3V448c0 17.7 14.3 32 32 32s32-14.3 32-32V146.3l32.4 35.4c11.9 13 32.2 13.9 45.2 2s13.9-32.2 2-45.2l-88-96zM320 480h32c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32zm0-128h96c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32zm0-128H480c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32zm0-128H544c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32z"
                fill="currentColor"
              />
            </svg>
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <div
      class="recharts-responsive-container"
      style="width: 100%; height: 300px; min-width: 0;"
    />
  </div>
</div>
`;

exports[`<SortableBarChartCard /> > renders properly ordered stats when ordering is set 3`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-up-wide-short ordering-dropdown__sort-icon"
              data-icon="arrow-up-wide-short"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M151.6 42.4C145.5 35.8 137 32 128 32s-17.5 3.8-23.6 10.4l-88 96c-11.9 13-11.1 33.3 2 45.2s33.3 11.1 45.2-2L96 146.3V448c0 17.7 14.3 32 32 32s32-14.3 32-32V146.3l32.4 35.4c11.9 13 32.2 13.9 45.2 2s13.9-32.2 2-45.2l-88-96zM320 480h32c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32zm0-128h96c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32zm0-128H480c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32zm0-128H544c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32z"
                fill="currentColor"
              />
            </svg>
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <div
      class="recharts-responsive-container"
      style="width: 100%; height: 300px; min-width: 0;"
    />
  </div>
</div>
`;

exports[`<SortableBarChartCard /> > renders properly ordered stats when ordering is set 4`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
          </button>
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-up-wide-short ordering-dropdown__sort-icon"
              data-icon="arrow-up-wide-short"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M151.6 42.4C145.5 35.8 137 32 128 32s-17.5 3.8-23.6 10.4l-88 96c-11.9 13-11.1 33.3 2 45.2s33.3 11.1 45.2-2L96 146.3V448c0 17.7 14.3 32 32 32s32-14.3 32-32V146.3l32.4 35.4c11.9 13 32.2 13.9 45.2 2s13.9-32.2 2-45.2l-88-96zM320 480h32c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32zm0-128h96c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32zm0-128H480c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32zm0-128H544c17.7 0 32-14.3 32-32s-14.3-32-32-32H320c-17.7 0-32 14.3-32 32s14.3 32 32 32z"
                fill="currentColor"
              />
            </svg>
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <div
      class="recharts-responsive-container"
      style="width: 100%; height: 300px; min-width: 0;"
    />
  </div>
</div>
`;

exports[`<SortableBarChartCard /> > renders properly paginated stats when pagination is set 1`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="disabled dropdown-item"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 me-3 dropdown-toggle btn btn-link"
          type="button"
        >
          Paginate
        </button>
        <div
          aria-hidden="true"
          class="dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              50
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              100
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              200
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              500
            </b>
             items per page
          </button>
          <hr
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear pagination
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <div
      class="recharts-responsive-container"
      style="width: 100%; height: 1100px; min-width: 0;"
    />
  </div>
  <div
    class="chart-card__footer--sticky card-footer"
  >
    <nav
      aria-label="pagination"
      class=""
    >
      <ul
        class="flex-wrap mb-0 simple-paginator justify-content-center pagination"
      >
        <li
          class="page-item disabled"
        >
          <span
            aria-label="Previous"
            class="page-link"
          >
            <span
              aria-hidden="true"
            >
              ‹
            </span>
            <span
              class="visually-hidden"
            >
              Previous
            </span>
          </span>
        </li>
        <li
          class="page-item active"
        >
          <span
            class="page-link"
            role="link"
          >
            1
          </span>
        </li>
        <li
          class="page-item"
        >
          <span
            class="page-link"
            role="link"
          >
            2
          </span>
        </li>
        <li
          class="page-item"
        >
          <span
            class="page-link"
            role="link"
          >
            3
          </span>
        </li>
        <li
          class="page-item"
        >
          <span
            class="page-link"
            role="link"
          >
            4
          </span>
        </li>
        <li
          class="page-item"
        >
          <span
            aria-label="Next"
            class="page-link"
          >
            <span
              aria-hidden="true"
            >
              ›
            </span>
            <span
              class="visually-hidden"
            >
              Next
            </span>
          </span>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`<SortableBarChartCard /> > renders properly paginated stats when pagination is set 2`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="disabled dropdown-item"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 me-3 dropdown-toggle btn btn-link"
          type="button"
        >
          Paginate
        </button>
        <div
          aria-hidden="true"
          class="dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              50
            </b>
             items per page
          </button>
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              100
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              200
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              500
            </b>
             items per page
          </button>
          <hr
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear pagination
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <div
      class="recharts-responsive-container"
      style="width: 100%; height: 2200px; min-width: 0;"
    />
  </div>
  <div
    class="chart-card__footer--sticky card-footer"
  >
    <nav
      aria-label="pagination"
      class=""
    >
      <ul
        class="flex-wrap mb-0 simple-paginator justify-content-center pagination"
      >
        <li
          class="page-item disabled"
        >
          <span
            aria-label="Previous"
            class="page-link"
          >
            <span
              aria-hidden="true"
            >
              ‹
            </span>
            <span
              class="visually-hidden"
            >
              Previous
            </span>
          </span>
        </li>
        <li
          class="page-item active"
        >
          <span
            class="page-link"
            role="link"
          >
            1
          </span>
        </li>
        <li
          class="page-item"
        >
          <span
            class="page-link"
            role="link"
          >
            2
          </span>
        </li>
        <li
          class="page-item"
        >
          <span
            aria-label="Next"
            class="page-link"
          >
            <span
              aria-hidden="true"
            >
              ›
            </span>
            <span
              class="visually-hidden"
            >
              Next
            </span>
          </span>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`<SortableBarChartCard /> > renders properly paginated stats when pagination is set 3`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="disabled dropdown-item"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 me-3 dropdown-toggle btn btn-link"
          type="button"
        >
          Paginate
        </button>
        <div
          aria-hidden="true"
          class="dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              50
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              100
            </b>
             items per page
          </button>
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              200
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              500
            </b>
             items per page
          </button>
          <hr
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear pagination
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <div
      class="recharts-responsive-container"
      style="width: 100%; height: 3520px; min-width: 0;"
    />
  </div>
</div>
`;

exports[`<SortableBarChartCard /> > renders properly paginated stats when pagination is set 4`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="disabled dropdown-item"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 me-3 dropdown-toggle btn btn-link"
          type="button"
        >
          Paginate
        </button>
        <div
          aria-hidden="true"
          class="dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              50
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              100
            </b>
             items per page
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              200
            </b>
             items per page
          </button>
          <button
            class="dropdown-item active"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <b>
              500
            </b>
             items per page
          </button>
          <hr
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <i>
              Clear pagination
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <div
      class="recharts-responsive-container"
      style="width: 100%; height: 3520px; min-width: 0;"
    />
  </div>
</div>
`;

exports[`<SortableBarChartCard /> > renders stats unchanged when no ordering is set 1`] = `
<div
  class="card"
  role="document"
>
  <div
    class="chart-card__header card-header"
  >
    Foo
    <div
      class="float-end"
    >
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="btn-sm p-0 dropdown-toggle btn btn-link"
          type="button"
        >
          Order by
        </button>
        <div
          aria-hidden="true"
          class="w-100 ordering-dropdown__menu--link dropdown-menu dropdown-menu-end"
          role="menu"
          tabindex="-1"
        >
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Name
          </button>
          <button
            class="dropdown-item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Amount
          </button>
          <div
            class="dropdown-divider"
            tabindex="-1"
          />
          <button
            class="disabled dropdown-item"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <i>
              Clear selection
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card-body"
  >
    <div
      class="recharts-responsive-container"
      style="width: 100%; height: 300px; min-width: 0;"
    />
  </div>
</div>
`;
